!!! TASK: Throw exceptions in unnecessary rebase and publish cases

mhsdesign opened the pull request:
based on https://github.com/neos/neos-development-collection/pull/5396

### new `WorkspaceCommandSkipped` exception:

Exception to denote that the workspace operation was skipped. No events were published.

*Workspace publishing/discarding*

Command skipped if there are no (selected) publishable changes.

To avoid this exception being thrown into the outer world it should be caught and logged or the case were its
likely that the operation might be a noop can easily be determined beforehand via `Workspace::hasPublishableChanges()`:

```php
if ($workspace->hasPublishableChanges()) {
    $contentRepository->handle(...);
}
```

*Workspace rebase*

Command skipped if the workspace is not outdated.

*Workspace base change*

Command skipped when attempting to change the base workspace to the currently set base workspace.

Note:

The case is not handled gracefully with a no-op as there would be no traces (emitted events) of the handled command,
and the original content stream id is kept (for publish operations).
This exception denoting the operation is obsolete should harden the interaction and make behaviour more explicit.


**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
!!! TASK: Throw exceptions in unnecessary rebase and publish cases

bwaidelich reviewed (dismissed):
You're a workhorse, thanks for taking care.
+1 by reading

mhsdesign commented:
It was @kitsunet work i downloaded the patch as i thought it would go to waste :D

bwaidelich commented:
> It was @kitsunet work i downloaded the patch as i thought it would go to waste :D

you are _both_ work horses 🐴 🐴

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Exception/NoChangesException.php:
```diff
@@ -0,0 +1,38 @@
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+
+/**
+ * @api thrown as part of command handling in case
```
```suggestion
 * @api thrown as part of command handling in case of empty publish and rebase operations
```

Fixing myself here lazy style

mhsdesign commented:
Okay well its not thaaaat easy ... 

Following case have to be thought out:


- what should happen if `nodesToPublish` is empty?
  - Should it even be allowed to be set to empty? Or enforced by the command that this CANNOT happen?
  - ... same for discard individual nodes... 
  - line: https://github.com/neos/neos-development-collection/blob/3e7035fe6c8c71bfac79c56c64358a0cc8f33207/Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php#L438


- what should happen if the node id in `nodesToPublish` did not match anything?
  - Eg either if the node never contained change
  - doesnt exist at all
  - or was concurrently published and doesnt contain changes anymore?
  - ... same for discard individual nodes... 
  - line: https://github.com/neos/neos-development-collection/blob/3e7035fe6c8c71bfac79c56c64358a0cc8f33207/Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php#L465


- what should happen on full or partial publish if there are no pending changes in the workspace?
  - line: https://github.com/neos/neos-development-collection/blob/3e7035fe6c8c71bfac79c56c64358a0cc8f33207/Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php#L182

- what should happen on full or partial **discard** if there are no pending changes in the workspace?
  - line: https://github.com/neos/neos-development-collection/blob/3e7035fe6c8c71bfac79c56c64358a0cc8f33207/Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php#L673


- what should happen on rebase if the workspace is not outdated? 
  - line: https://github.com/neos/neos-development-collection/blob/3e7035fe6c8c71bfac79c56c64358a0cc8f33207/Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php#L347

kitsunet commented:
> Okay well its not thaaaat easy ...
> 
> Following case have to be thought out:
> 
>     * what should happen if `nodesToPublish` is empty?

I think we could/should prevent this in the command?! If we want to throw we can decide further up front (aka UI).

>     * what should happen if the node id in `nodesToPublish` did not match anything?

IMHO that is a fair possibility that can happen with race conditions on shared workspaces for example, so I would handle it gracefully and accept this as no-op. BUT given next point I probably opt for exception and graceful handling further up the chain.

>     * what should happen on full or partial publish if there are no pending changes in the workspace?

Yeah this is the tricky case compared to the above, IF you had a race condition and send identifiers to a partial publish that are no longer changed this is similar to starting a full publish if there are no pending changes (anymore).

>     * what should happen on full or partial **discard** if there are no pending changes in the workspace?

Same as above I guess. If we start to throw.

>     * what should happen on rebase if the workspace is not outdated?

Probably again same as above. 



For all of this though we should make sure there is a way out even for unexpected situations. Which could e.g. be the rebase or discard even if things look unchanged.

mhsdesign commented:
Todo rebase on https://github.com/neos/neos-development-collection/pull/5385
and in the constructor of partial publish and discard require the node ids to be NOT empty.

As the $nodeIdsToPublish is rather to be seen as filter than as expectation - imagine you enter a non-existing node aggregate id - id say that its okay if this is then a noop? one cannot be mad if a filter doesnt match anything? Also what if one of the nodes in the filter is unmatched but there is one node found to be published? If we throw it would only be consistent to make it required that all node ids specified MUST translate to a change.

That is somehow not really the behaviour a filter is made for. Thought especially throwing would at least give feed back to users in the Ui instead that the publish button keeps being orange like shown here: https://github.com/neos/neos-development-collection/issues/4614
With the simplified publishing that cannot happen anymore but yeah ... im not fully sure.

mhsdesign commented:
Todo: Also other noops like in change base workspace are not desired and have to be prevented in `\Neos\Neos\Domain\Service\WorkspacePublishingService::changeBaseWorkspace` instead!

mhsdesign committed:
TASK: Prevent publish/discard commands to exist with empty filter

mhsdesign committed:
TASK: Throw NoChangesException consistently also during discard and adjust tests

mhsdesign committed:
PATCH: Restore workspace rebase and throw skipped exception instead of noop

mhsdesign committed:
TASK: Introduce BaseWorkspaceUnchangedException to avoid no-op for ChangeBaseWorkspace

mhsdesign commented:
Okay i polished the change and adjusted the tests.

Also i checked in the command handling https://github.com/neos/neos-development-collection/blob/47fc20b6487c8038f41f66d28f1955b98e1b5e72/Neos.ContentRepository.Core/Classes/ContentRepository.php#L110 that when we use `yield` that we never return and no-op without issuing any events. This is basically the followup of the philosophy from https://github.com/neos/neos-development-collection/pull/5357 and https://github.com/neos/neos-development-collection/pull/5396 to ensure that each command maps to at least one event.

mhsdesign committed:
TASK: Improve documentation of `WorkspaceRebaseFailed` exception

mhsdesign committed:
TASK: Document possible occurrence of NoChangesException in WorkspacePublishingService

kitsunet commented:
The behavior test errors are valid as the expectation changed now (we gracefully handled the empty rebase before).

Unlike what I wrote in December, I am pretty convinced now that we are making a mistake with all this, but as I wrote in another place, we can revert from this stance easier than the other way around so I am ok with merging it given the majority seems to be in favor of doing it this way.

kitsunet reviewed (approved):
Looks fine, test error still left to fix but that should be easy enough.

mhsdesign committed:
TASK: Fix asset usage tests by removing obsolete rebase

mhsdesign committed:
TASK: Fix cr security tests

separate more complex discard commands into own tests as attempting to discard 0 nodes fails now and its harder to test that way in batches.

mhsdesign committed:
TASK: Rename NoChangesException to WorkspaceCommandSkipped

Instead of keeping it specific to publishing and naming it `NoPublishableChanges` we use it as generic no-op exception for workspace commands. Especially because `becauseWorkspaceToRebaseIsNotOutdated` already deviates from the 'PublishableChanges' kind.

mhsdesign committed:
TASK: Combine `BaseWorkspaceUnchangedException` with `WorkspaceCommandSkipped`

mhsdesign commented:
Fyi i also followed your idea to have one exception for these things instead of the `NoChangesException` (which i interpreted as no _publishable_ changes) and `BaseWorkspaceUnchangedException`. Its not all part of the `WorkspaceCommandSkipped` which makes this behaviour now even more clear.
